src/peft/tuners/lora.py

```diff
@@ -1146,3 +1176,55 @@ def forward(self, x: torch.Tensor):
                         )
                     result += output
                 return result
+
+
+class QuantLinear(torch.nn.Module, LoraLayer):
+    def __init__(
+        self,
+        adapter_name,
+        quant_linear_module,
+        r: int = 0,
+        lora_alpha: int = 1,
+        lora_dropout: float = 0.0,
+        **kwargs,
+    ):
+        torch.nn.Module.__init__(self)
+        LoraLayer.__init__(
+            self, in_features=quant_linear_module.infeatures, out_features=quant_linear_module.outfeatures
+        )
+        self.quant_linear_module = quant_linear_module
```

Comment:
This is an interesting deviation from how the other lora layers are implemented. Here, we pass the original layer (`quant_linear_module`) and use it under the hood. For the normal `Linear` lora layer, we don't get the layer, instead basically creating a new linear layer:

```
# in __init__
nn.Linear.__init__(self, in_features, out_features, **kwargs)
# in forward
result = F.linear(x, transpose(self.weight, self.fan_in_fan_out), bias=self.bias)
```

I actually prefer the solution here but wonder if there was a specific reason why this approach was _not_ taken originally. If so, would that same reason apply here or are we good with having two different approaches? Hopefully, the others can clarify this.

Reply:
I did that because I wanted to put the new QuantLinear class in the same place as the others for conformity. If we want to go with the same approach, we will need to create this new class in a function so that the `auto_gptq` import is protected ( to avoid circular import as it is also importing peft ) . LMK what you think about this solution.